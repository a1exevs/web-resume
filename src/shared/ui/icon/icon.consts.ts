export const ICON_DEFAULT_COLOR = 'currentColor';
