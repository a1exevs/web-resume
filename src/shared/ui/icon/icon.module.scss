@import 'src/shared';

:host {
  font-size: 0;
  line-height: 0;
}

.Icon {
  &_appFontSize {
    width: 1rem;
    height: 1rem;
  }
  &_parentFontSize {
    width: 1em;
    height: 1em;
  }
  &_smallSize {
    width: 8px;
    height: 8px;
  }
  &_normalSize {
    width: 3 * 8px;
    height: 3 * 8px;
  }

  &_mediumSize {
    width: 4 * 8px;
    height: 4 * 8px;
  }
  &_LargeSize {
    width: 8 * 8px;
    height: 8 * 8px;
  }
  &_superGiantSize {
    width: 16 * 8px;
    height: 16 * 8px;
  }
}
