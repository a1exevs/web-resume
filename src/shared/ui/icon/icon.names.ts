/**
 * For adding a new svg-icon
 * 1) convert svg to symbol manually or with using online-converters
 *   (for example, https://svg-to-symbol-app.pages.dev)
 *   @example
 *   SVG:
 *   <svg
 *     xmlns="http://www.w3.org/2000/svg"
 *     width="10"
 *     height="10"
 *   >
 *     <rect
 *       width="10"
 *       height="10"
 *       fill="green"
 *     />
 *   </svg>
 *
 *   Symbol:
 *   <symbol
 *     id="svg-b0dfq0d60"
 *     viewBox="0 0 10 10"
 *     xmlns="http://www.w3.org/2000/svg"
 *   >
 *     <path
 *       fill="green"
 *       d="M0 0h10v10H0z"
 *     />
 *   </symbol>
 * 2) remove extra attributes such as xmlns
 * 3) replace fill/color attribute value with 'currentColor'
 * 4) use your unique and simple id. You need to add this id in IconName
 * 5) add new symbol in icons.svg (src/assets/icons/icons.svg)
 *   @example
 *   Symbol:
 *   <symbol
 *     id="rect"
 *     viewBox="0 0 10 10"
 *   >
 *     <path
 *       fill="currentColor"
 *       d="M0 0h10v10H0z"
 *     />
 *   </symbol>
 *   @see IconName
 */
export enum IconName {
  RECT = 'rect',
  GITHUB = 'github',
  HEAD_HUNTER = 'head-hunter',
  LINKEDIN = 'linkedin',
  EMAIL = 'email',
  YOUTUBE = 'youtube',
  LEETCODE = 'leetcode',
  INSTAGRAM = 'instagram',
  WHATSAPP = 'whatsapp',
  TELEGRAM = 'telegram',
}
