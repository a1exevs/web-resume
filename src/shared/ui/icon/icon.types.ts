export enum IconSize {
  APP_FONT_SIZE = 'AppFontSize',
  PARENT_FONT_SIZE = 'ParentFontSize',
  SMALL = 'Small',
  NORMAL = 'Normal',
  MEDIUM = 'Medium',
  LARGE = 'Large',
  SUPER_GIANT = 'SuperGiant',
}
