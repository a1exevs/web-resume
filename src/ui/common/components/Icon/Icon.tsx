import React from 'react';

export enum IconName {
  TECHNICAL_SHILLS = 'TECHNICAL_SHILLS',
  CAREER_HISTORY = 'CAREER_HISTORY',
  PERSONAL_INFORMATION = 'PERSONAL_INFORMATION',
}

const getIconPath = (iconName: IconName): React.JSX.Element => {
  switch (iconName) {
    case IconName.TECHNICAL_SHILLS: {
      return (
        <path d="M12 3 1 9l4 2.18v6L12 21l7-3.82v-6l2-1.09V17h2V9L12 3zm6.82 6L12 12.72 5.18 9 12 5.28 18.82 9zM17 15.99l-5 2.73-5-2.73v-3.72L12 15l5-2.73v3.72z"></path>
      );
    }
    case IconName.CAREER_HISTORY: {
      return <path d="M12 7.77 18.39 18H5.61L12 7.77M12 4 2 20h20L12 4z"></path>;
    }
    case IconName.PERSONAL_INFORMATION: {
      return (
        <path d="M11 23.59v-3.6c-5.01-.26-9-4.42-9-9.49C2 5.26 6.26 1 11.5 1S21 5.26 21 10.5c0 4.95-3.44 9.93-8.57 12.4l-1.43.69zM11.5 3C7.36 3 4 6.36 4 10.5S7.36 18 11.5 18H13v2.3c3.64-2.3 6-6.08 6-9.8C19 6.36 15.64 3 11.5 3zm-1 11.5h2v2h-2zm2-1.5h-2c0-3.25 3-3 3-5 0-1.1-.9-2-2-2s-2 .9-2 2h-2c0-2.21 1.79-4 4-4s4 1.79 4 4c0 2.5-3 2.75-3 5z"></path>
      );
    }
    default: {
      return <div></div>;
    }
  }
};

type IconProps = {
  name: IconName;
  width?: string;
  height?: string;
  color?: string;
};

export const Icon: React.FC<IconProps> = ({
  name,
  width = '1.5em',
  height = '1.5em',
  color = 'currentColor',
  ...rest
}) => {
  return (
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" width={width} height={height} fill={color} {...rest}>
      {getIconPath(name)}
    </svg>
  );
};
