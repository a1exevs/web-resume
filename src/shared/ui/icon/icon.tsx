import { isEmpty, isUndefined } from '@alexevs/ts-guards';
import cn from 'classnames';
import React from 'react';

import icons from 'src/assets/icons/icons.svg';
import { ICON_DEFAULT_COLOR } from 'src/shared/ui/icon/icon.consts';
import classes from 'src/shared/ui/icon/icon.module.scss';
import { IconName } from 'src/shared/ui/icon/icon.names';
import { IconSize } from 'src/shared/ui/icon/icon.types';

type Props = {
  size?: IconSize;
  color?: string;
  width?: string;
  height?: string;
  className?: string;
  title?: string;
  icon: IconName;
};

const IconSizesMap: Record<IconSize, string> = {
  [IconSize.APP_FONT_SIZE]: classes.Icon_appFontSize,
  [IconSize.PARENT_FONT_SIZE]: classes.Icon_parentFontSize,
  [IconSize.SMALL]: classes.Icon_smallSize,
  [IconSize.NORMAL]: classes.Icon_normalSize,
  [IconSize.MEDIUM]: classes.Icon_mediumSize,
  [IconSize.LARGE]: classes.Icon_LargeSize,
  [IconSize.SUPER_GIANT]: classes.Icon_superGiantSize,
};

const Icon: React.FC<Props> = ({
  size = IconSize.APP_FONT_SIZE,
  icon,
  color = ICON_DEFAULT_COLOR,
  width,
  height,
  className = '',
  title,
}) => {
  const getSizeClass = (sizeMap: Record<IconSize, string>): string => {
    if (!isUndefined(width) || !isUndefined(height)) {
      return '';
    }
    return sizeMap[size];
  };

  return (
    <svg
      className={cn(classes.Icon, getSizeClass(IconSizesMap), className)}
      fill={color}
      color={color}
      width={width}
      height={height}
    >
      {!isEmpty(title) && <title>{title}</title>}
      <use xlinkHref={`${icons}#${icon}`}></use>
    </svg>
  );
};

export default Icon;
