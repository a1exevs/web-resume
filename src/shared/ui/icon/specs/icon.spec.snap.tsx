import { render } from '@testing-library/react';
import React from 'react';

import Icon from 'src/shared/ui/icon/icon';
import { IconName } from 'src/shared/ui/icon/icon.names';
import { IconSize } from 'src/shared/ui/icon/icon.types';

describe('Icon component', () => {
  it('default rendering.', () => {
    const { asFragment } = render(<Icon icon={IconName.RECT} />);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('"color" prop', () => {
    it('prop "color" has a value', () => {
      const { asFragment } = render(<Icon icon={IconName.RECT} color="red" />);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('"size" prop', () => {
    it('"size" prop has value = "Small"', () => {
      const { asFragment } = render(<Icon icon={IconName.RECT} size={IconSize.SMALL} />);
      expect(asFragment()).toMatchSnapshot();
    });
    it('"size" prop has value = "Normal"', () => {
      const { asFragment } = render(<Icon icon={IconName.RECT} size={IconSize.NORMAL} />);
      expect(asFragment()).toMatchSnapshot();
    });
    it('"size" prop has value = "Large"', () => {
      const { asFragment } = render(<Icon icon={IconName.RECT} size={IconSize.LARGE} />);
      expect(asFragment()).toMatchSnapshot();
    });
    it('"size" prop has value = "SuperGiant"', () => {
      const { asFragment } = render(<Icon icon={IconName.RECT} size={IconSize.SUPER_GIANT} />);
      expect(asFragment()).toMatchSnapshot();
    });
    it('"size" prop has value = "AppFontSize"', () => {
      const { asFragment } = render(<Icon icon={IconName.RECT} size={IconSize.APP_FONT_SIZE} />);
      expect(asFragment()).toMatchSnapshot();
    });
    it('"size" prop has value = "ParentFontSize"', () => {
      const { asFragment } = render(<Icon icon={IconName.RECT} size={IconSize.PARENT_FONT_SIZE} />);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('"width" and "height" props', () => {
    it('"width" and "height" props have some values', () => {
      const { asFragment } = render(<Icon icon={IconName.RECT} width="'24px'" height="24px" />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
