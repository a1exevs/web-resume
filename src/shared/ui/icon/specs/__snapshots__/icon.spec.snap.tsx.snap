// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Icon component "color" prop prop "color" has a value 1`] = `
<DocumentFragment>
  <svg
    class="Icon Icon_appFontSize"
    color="red"
    fill="red"
  >
    <use
      xlink:href="FileMock#rect"
    />
  </svg>
</DocumentFragment>
`;

exports[`Icon component "size" prop "size" prop has value = "AppFontSize" 1`] = `
<DocumentFragment>
  <svg
    class="Icon Icon_appFontSize"
    color="currentColor"
    fill="currentColor"
  >
    <use
      xlink:href="FileMock#rect"
    />
  </svg>
</DocumentFragment>
`;

exports[`Icon component "size" prop "size" prop has value = "Large" 1`] = `
<DocumentFragment>
  <svg
    class="Icon Icon_LargeSize"
    color="currentColor"
    fill="currentColor"
  >
    <use
      xlink:href="FileMock#rect"
    />
  </svg>
</DocumentFragment>
`;

exports[`Icon component "size" prop "size" prop has value = "Normal" 1`] = `
<DocumentFragment>
  <svg
    class="Icon Icon_normalSize"
    color="currentColor"
    fill="currentColor"
  >
    <use
      xlink:href="FileMock#rect"
    />
  </svg>
</DocumentFragment>
`;

exports[`Icon component "size" prop "size" prop has value = "ParentFontSize" 1`] = `
<DocumentFragment>
  <svg
    class="Icon Icon_parentFontSize"
    color="currentColor"
    fill="currentColor"
  >
    <use
      xlink:href="FileMock#rect"
    />
  </svg>
</DocumentFragment>
`;

exports[`Icon component "size" prop "size" prop has value = "Small" 1`] = `
<DocumentFragment>
  <svg
    class="Icon Icon_smallSize"
    color="currentColor"
    fill="currentColor"
  >
    <use
      xlink:href="FileMock#rect"
    />
  </svg>
</DocumentFragment>
`;

exports[`Icon component "size" prop "size" prop has value = "SuperGiant" 1`] = `
<DocumentFragment>
  <svg
    class="Icon Icon_superGiantSize"
    color="currentColor"
    fill="currentColor"
  >
    <use
      xlink:href="FileMock#rect"
    />
  </svg>
</DocumentFragment>
`;

exports[`Icon component "width" and "height" props "width" and "height" props have some values 1`] = `
<DocumentFragment>
  <svg
    class="Icon"
    color="currentColor"
    fill="currentColor"
    height="24px"
    width="'24px'"
  >
    <use
      xlink:href="FileMock#rect"
    />
  </svg>
</DocumentFragment>
`;

exports[`Icon component default rendering. 1`] = `
<DocumentFragment>
  <svg
    class="Icon Icon_appFontSize"
    color="currentColor"
    fill="currentColor"
  >
    <use
      xlink:href="FileMock#rect"
    />
  </svg>
</DocumentFragment>
`;
